common/src/main/java/com/google/auto/common/BasicAnnotationProcessor.java

```diff
@@ -196,52 +204,54 @@ public final boolean process(Set<? extends TypeElement> annotations, RoundEnviro
       postRound(roundEnv);
       if (!roundEnv.errorRaised()) {
         reportMissingElements(
-            ImmutableSet.<ElementName>builder()
-                .addAll(deferredElementNames)
+            ImmutableSet.<ElementFactory>builder()
+                .addAll(deferredElementFactories)
                 .addAll(elementsDeferredBySteps.values())
                 .build());
       }
       return false;
     }
 
-    process(validElements(roundEnv));
+    process(getWellFormedElementsByAnnotationType(roundEnv));
 
     postRound(roundEnv);
 
     return false;
   }
 
   /** Processes the valid elements, including those previously deferred by each step. */
-  private void process(ImmutableSetMultimap<TypeElement, Element> validElements) {
+  private void process(ImmutableSetMultimap<TypeElement, Element> wellFormedElements) {
     for (Step step : steps) {
       ImmutableSet<TypeElement> annotationTypes = getSupportedAnnotationTypeElements(step);
       ImmutableSetMultimap<TypeElement, Element> stepElements =
           new ImmutableSetMultimap.Builder<TypeElement, Element>()
               .putAll(indexByAnnotation(elementsDeferredBySteps.get(step), annotationTypes))
-              .putAll(filterKeys(validElements, Predicates.in(annotationTypes)))
+              .putAll(filterKeys(wellFormedElements, annotationTypes::contains))
               .build();
       if (stepElements.isEmpty()) {
         elementsDeferredBySteps.removeAll(step);
       } else {
         Set<? extends Element> rejectedElements =
             step.process(toClassNameKeyedMultimap(stepElements));
         elementsDeferredBySteps.replaceValues(
-            step, transform(rejectedElements, ElementName::forAnnotatedElement));
+            step,
+            rejectedElements.stream()
+                .map(element -> ElementFactory.forAnnotatedElement(element, messager))
+                .collect(toImmutableList()));
       }
     }
   }
 
-  private void reportMissingElements(Set<ElementName> missingElementNames) {
-    for (ElementName missingElementName : missingElementNames) {
-      Optional<? extends Element> missingElement = missingElementName.getElement(elements);
-      if (missingElement.isPresent()) {
+  private void reportMissingElements(Set<ElementFactory> missingElementFactories) {
+    for (ElementFactory missingElementFactory : missingElementFactories) {
+      Element missingElement = missingElementFactory.getElement(elementUtils);
+      if (missingElement != null) {
```

Comment:
Since empty is expected, I personally prefer the use of `Optional` instead.

Reply:
I would be inclined to agree if we could use `Optional.ifPresentOrElse` here. Then we wouldn't have the ugly `.ifPresent...get` pattern. Unfortunately we're still building on Java 8 and `ifPresentOrElse` was only added in Java 9.